crates/ruff_python_parser/src/parser/mod.rs

```diff
@@ -0,0 +1,4427 @@
+use std::fmt::Display;
+
+use self::helpers::token_kind_to_cmp_op;
+use crate::{
+    error::FStringErrorType,
+    lexer::{LexResult, LexicalError, Spanned},
+    string::{
+        concatenated_strings, parse_fstring_literal_element, parse_string_literal, StringType,
+    },
+    token_set::TokenSet,
+    Mode, ParseError, ParseErrorType, Tok, TokenKind,
+};
+use ast::{
+    BoolOp, CmpOp, ConversionFlag, ExceptHandler, ExprContext, FStringElement, IpyEscapeKind, Mod,
+    Number, Operator, Pattern, Singleton, UnaryOp,
+};
+use bitflags::bitflags;
+use itertools::PeekNth;
+use ruff_python_ast::{self as ast, Expr, Stmt};
+use ruff_text_size::{Ranged, TextLen, TextRange, TextSize};
+
+mod helpers;
+mod tests;
+
+pub(crate) fn parse_ok_tokens(
+    lxr: impl IntoIterator<Item = Spanned>,
+    source: &str,
+    mode: Mode,
+) -> Result<Mod, ParseError> {
+    let lxr = lxr
+        .into_iter()
+        .filter(|(tok, _)| !matches!(tok, Tok::Comment { .. } | Tok::NonLogicalNewline))
+        .map(Ok::<(Tok, TextRange), LexicalError>);
+    let parsed_file = Parser::new(source, mode, lxr).parse();
+    if parsed_file.parse_errors.is_empty() {
+        Ok(parsed_file.ast)
+    } else {
+        Err(parsed_file.parse_errors.into_iter().next().unwrap())
+    }
+}
+
+#[derive(Debug)]
+pub struct ParsedFile {
+    pub ast: ast::Mod,
+    pub parse_errors: Vec<ParseError>,
+}
+
+bitflags! {
+    #[derive(Default, Debug, Clone, Copy, PartialEq, Eq)]
+    struct ParserCtxFlags: u16 {
+        const TUPLE_EXPR = 1 << 0;
+        const PARENTHESIZED_EXPR = 1 << 1;
+        const BRACKETSIZED_EXPR = 1 << 2;
+        const BRACESIZED_EXPR = 1 << 3;
+        const LAMBDA_EXPR = 1 << 4;
+
+        const IF_STMT = 1 << 5;
+        const FUNC_DEF_STMT = 1 << 6;
+        const CLASS_DEF_STMT = 1 << 7;
+        const WITH_STMT = 1 << 8;
+        const FOR_STMT = 1 << 9;
+        const WHILE_STMT  = 1 << 10;
+        const MATCH_STMT  = 1 << 11;
+
+        const ARGUMENTS = 1 << 12;
+        const FOR_TARGET = 1 << 13;
+    }
+}
+
+type ExprWithRange = (Expr, TextRange);
+type StmtWithRange = (Stmt, TextRange);
+
+/// Binding power associativity
+enum Associativity {
+    Left,
+    Right,
+}
+
+pub(crate) struct Parser<'src, I>
+where
+    I: Iterator<Item = LexResult>,
+{
+    source: &'src str,
+    lexer: PeekNth<I>,
+    /// Stores all the syntax errors found during the parsing.
+    errors: Vec<ParseError>,
+    /// This tracks the current expression or statement being parsed. For example,
+    /// if we're parsing a tuple expression, e.g. `(1, 2)`, `ctx` has the value
+    /// `ParserCtxFlags::TUPLE_EXPR`.
+    ///
+    /// The `ctx` is also used to create custom error messages and forbid certain
+    /// expressions or statements of being parsed. The `ctx` should be empty after
+    /// an expression or statement is done parsing.
+    ctx: ParserCtxFlags,
```

Comment:
I think it should be possible to remove most usages of `ctx` by instead pass the context to the specific parsing function:

* `parse_body`: Create a `Clause` enum with variants for each possible clause that can have a body. Add an explicit `parent_clause` argument to `parse_body`. 
* `parse_paremters`: Create a dedicated `FunctionKind` enum with a variant for `Lambda` and `FunctionDefinition`. Explicilty pass the `function_kind` to `parse_parameters`.

Removing the need for `ctx_stack` is probably the trickiest but we can do it by using the call-stack by:

* Copying the existing `ctx` flags before changing them
* Changing the context flags
* Call into the sub-parsing
* Restoring the context flags afterwards

This won't give you access to the parent context while parsing, but I wonder if that's even necessary. All we need to know is if we're inside of a for target. We can keep that context flag set for the entire target expression (and we can explicilty set it back to `false` if there's an expression where it needs to be restored to false).

Reply:
I wonder if we could remove `last_ctx` by changing `parse_expr` to return a `ParsedExpression` struct that has two fields:

* The parsed expression
* A boolean field whether the expression was parenthesized

Reply:
That might work.

Reply:
> I wonder if we could remove `last_ctx` by changing `parse_expr` to return a `ParsedExpression` struct that has two fields:
> 
>     * The parsed expression
> 
>     * A boolean field whether the expression was parenthesized

To be able to remove `last_ctx` we need to create another solution for that problem in `parse_with_items`.